x/superfluid/keeper/migrate_test.go

```diff
@@ -0,0 +1,224 @@
+			superfluidKeeper := suite.App.SuperfluidKeeper
+			lockupKeeper := suite.App.LockupKeeper
+			stakingKeeper := suite.App.StakingKeeper
+			poolmanagerKeeper := suite.App.PoolManagerKeeper
+
+			// Generate and fund two accounts.
+			// Account 1 will be the account that creates the pool.
+			// Account 2 will be the account that joins the pool.
+			delAddrs := CreateRandomAccounts(2)
+			poolCreateAcc := delAddrs[0]
+			poolJoinAcc := delAddrs[1]
+			for _, acc := range delAddrs {
+				err := simapp.FundAccount(bankKeeper, ctx, acc, defaultAcctFunds)
+				suite.Require().NoError(err)
+			}
+
+			// Set up a single validator.
+			valAddr := suite.SetupValidator(stakingtypes.BondStatus(stakingtypes.Bonded))
+
+			// Create a balancer pool of "stake" and "foo".
+			msg := balancer.NewMsgCreateBalancerPool(poolCreateAcc, balancer.PoolParams{
+				SwapFee: sdk.NewDecWithPrec(1, 2),
+				ExitFee: sdk.NewDec(0),
+			}, defaultPoolAssets, defaultFutureGovernor)
+			balancerPooId, err := poolmanagerKeeper.CreatePool(ctx, msg)
+			suite.Require().NoError(err)
+
+			// Join the balancer pool.
+			// Note the account balance before and after joining the pool.
+			balanceBeforeJoin := bankKeeper.GetAllBalances(ctx, poolJoinAcc)
+			_, _, err = gammKeeper.JoinPoolNoSwap(ctx, poolJoinAcc, balancerPooId, gammtypes.OneShare.MulRaw(50), sdk.Coins{})
+			suite.Require().NoError(err)
+			balanceAfterJoin := bankKeeper.GetAllBalances(ctx, poolJoinAcc)
+
+			// The balancer join pool amount is the difference between the account balance before and after joining the pool.
+			joinPoolAmt, _ := balanceBeforeJoin.SafeSub(balanceAfterJoin)
+
+			// Determine the pool's LP token denomination.
+			balancerPool, err := gammKeeper.GetPoolAndPoke(ctx, balancerPooId)
+			suite.Require().NoError(err)
+			poolDenom := gammtypes.GetPoolShareDenom(balancerPool.GetId())
+
+			// Register the LP token as a superfluid asset
+			err = superfluidKeeper.AddNewSuperfluidAsset(ctx, types.SuperfluidAsset{
+				Denom:     poolDenom,
+				AssetType: types.SuperfluidAssetTypeLPShare,
+			})
+			suite.Require().NoError(err)
+
+			// Note how much of the LP token the account that joined the pool has.
+			poolShareOut := bankKeeper.GetBalance(ctx, poolJoinAcc, poolDenom)
+
+			// Create a cl pool with the same underlying assets as the balancer pool.
+			clPool := suite.PrepareCustomConcentratedPool(poolCreateAcc, defaultPoolAssets[0].Token.Denom, defaultPoolAssets[1].Token.Denom, 1, sdk.NewInt(-6), sdk.ZeroDec())
+
+			// Add a sanctioned link between the balancer and concentrated liquidity pool.
+			migrationRecord := gammtypes.MigrationRecords{BalancerToConcentratedPoolLinks: []gammtypes.BalancerToConcentratedPoolLink{
+				{BalancerPoolId: balancerPool.GetId(), ClPoolId: clPool.GetId()},
+			}}
+			gammKeeper.SetMigrationInfo(ctx, migrationRecord)
+
+			// The amount of coins we lock is the amount of LP tokens the account that joined the pool has.
+			coinsToLock := poolShareOut
```

Comment:
Let's add a test where lock has a different number of shares locked from what we are trying to migrate

Reply:
Great point, I modified all the tests here to use varying percentages of the initial join amount! https://github.com/osmosis-labs/osmosis/pull/4129/commits/d961640126dbf52f7b779ae531a710723ecb92e3